### ISSUE TITLE
Feature gate icons

### ISSUE THREAD

**Issue by Stefanuk12** [carloskiki/leptos-icons#46]
As you have acknowledged, having so many icons causes rust-analyzer to lag. To fix this, you can gate each icon into its own feature flag. If you reach over the maximum features, you can group icons instead and do it that way.

**Comment by carloskiki** [comment]
Yes, this crate is the cause which instigated limits on crate features: https://blog.rust-lang.org/2023/10/26/broken-badges-and-23k-keywords.html.

The only obvious way I see of partitioning icons with feature flags would be on a _per-library_ basis. This feels somewhat of a coarse grain method, but I could get behind it as it will improve UX.

If you have time, I would encourage you to open a PR and try your hand at implementing _per-library_ features. I would be happy to review it :), but I do not have the time to implement it myself.

**Comment by martijnarts** [comment]
This would have to happen on the https://github.com/carloskiki/icondata/ repo, right?

**Comment by carloskiki** [comment]
Yes exactly!

For a bit of context, what we did before is that we feature gated every single icon individually. This created a lot of needless churn for the compiler and Rust's infrastructure. I refer you to [this blog post](https://blog.rust-lang.org/2023/10/26/broken-badges-and-23k-keywords.html) from the official Rust team which describes the issue in detail.

Following this, we decided not to use any feature flags, the only downside being that now `rust-analyzer` is under a lot of pressure from our crate (due to the massive amount of `static`s we make available).

I think a nice granularity to have would be icon-lib-wise; i.e., Font Awesome, Octicons, etc., would each have a feature flag. This would leave us with 20 or so feature flags. This would alleviate some of the work for RA, and would not be to painful for the end user.
What do you think?

**Comment by martijnarts** [comment]
I'd question what the purpose is of feature flagging. If it's only to reduce strain on RA, I think doing one flag per icon lib is perfectly fine!

I don't think having more icons increases something like the final binary size, right, since they'd just get compiled out? So then having per-icon flags doesn't really do much, I suppose?

**Comment by martijnarts** [comment]
I've opened a PR for the feature flags here: https://github.com/carloskiki/icondata/pull/39 :)

**Comment by axos88** [comment]
Are unused icons optimized away from the generated wasm binary? Or is the entire icon set written to the final binary? I'm not entirely sure whether that's automatic, or something needs to be done? Obviously wouldn't want 23k icons to bloat my binary.

**Comment by carloskiki** [comment]
> Are unused icons optimized away from the generated wasm binary? Or is the entire icon set written to the final binary? I'm not entirely sure whether that's automatic, or something needs to be done? Obviously wouldn't want 23k icons to bloat my binary.

No, by personal experience all unused icons are removed by Dead Code Elimination. I've also never seen anyone show that icons were kept, although we are indeed relying on the compiler to do the work for us. You can (should) always inspect your `wasm` bundle with [`twiggy`](https://github.com/rustwasm/twiggy).
